twoliter/embedded/imghelper

```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+  local var val
+  for var in AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN; do
+    val="${var,,}"
+    val="${HOME}/.aws/${val//_/-}.env"
+    [[ -s "${val}" ]] || continue
+    declare -x "${var}=$(cat "${val}")"
+  done
+  # Verify that AWS credentials are functional.
+  aws sts get-caller-identity
+  # Log all PKCS11 helper activity, to simplify debugging.
+  export AWS_KMS_PKCS11_DEBUG=1
+  shim_sign_key=(-c shim-sign-key -t shim-sign-key)
+  code_sign_key=(-c code-sign-key -t code-sign-key)
+
+  sbsetup_wrapup aws
+}
+
+sbsetup_local_profile() {
+  local -n shim_sign_key code_sign_key
+  shim_sign_key="${1:?}"
+  code_sign_key="${2:?}"
+
+  # Disable the PKCS11 helper.
+  rm /etc/pkcs11/modules/aws-kms-pkcs11.module
+
+  # Generate the PKCS12 archives for import.
+  openssl pkcs12 \
+    -export \
+    -passout pass: \
+    -inkey "${SBKEYS}/shim-sign.key" \
+    -in "${SBKEYS}/shim-sign.crt" \
+    -certfile "${SBKEYS}/db.crt" \
+    -out "${SBKEYS}/shim-sign.p12"
+
+  openssl pkcs12 \
+    -export \
+    -passout pass: \
+    -inkey "${SBKEYS}/code-sign.key" \
+    -in "${SBKEYS}/code-sign.crt" \
+    -certfile "${SBKEYS}/vendor.crt" \
+    -out "${SBKEYS}/code-sign.p12"
+
+  # Import certificates and private key archive.
+  local pedb="/etc/pki/pesign"
+
+  certutil -d "${pedb}" -A -n db -i "${SBKEYS}/db.crt" -t ",,C"
+  certutil -d "${pedb}" -A -n shim-sign-key -i "${SBKEYS}/shim-sign.crt" -t ",,P"
+  pk12util -d "${pedb}" -i "${SBKEYS}/shim-sign.p12" -W ""
+
+  certutil -d "${pedb}" -A -n vendor -i "${SBKEYS}/vendor.crt" -t ",,C"
+  certutil -d "${pedb}" -A -n code-sign-key -i "${SBKEYS}/code-sign.crt" -t ",,P"
+  pk12util -d "${pedb}" -i "${SBKEYS}/code-sign.p12" -W ""
+
+  certutil -d "${pedb}" -L
+  shim_sign_key=(-c shim-sign-key)
+  code_sign_key=(-c code-sign-key)
+
+  sbsetup_wrapup local
+}
+
+uefi_pesign() {
+  local efi_mount arch shims shim grubs grub mokms mokm
+  local -n shim_sign_key code_sign_key
+  efi_mount="${1:?}"
+  arch="${2:?}"
+  shim_sign_key="${3:?}"
+  code_sign_key="${4:?}"
+
+  pushd "${efi_mount}/EFI/BOOT" >/dev/null || exit 1
+
+  shims=(boot*.efi)
+  shim="${shims[0]}"
+  grubs=(grub*.efi)
+  grub="${grubs[0]}"
+  mokms=(mm*.efi)
+  mokm="${mokms[0]}"
+
+  # Convert the vendor certificate to the expected format.
+  cert_table "${SBKEYS}/vendor.cer" "${SBKEYS}/vendor.obj"
+
+  # Replace the embedded vendor certificate, then sign shim with the db key.
+  if [[ "$(pesign -i "${shim}" -l)" == "No signatures found." ]]; then
+    mv "${shim}" "${shim}.unsigned"
+  else
+    pesign -i "${shim}" -o "${shim}.unsigned" -u 0 -r
+  fi
+  "${arch}-objdumpcopy" "${shim}.unsigned" \
+    --update-section ".vendor_cert=${SBKEYS}/vendor.obj"
+  pesign -i "${shim}.unsigned" -o "${shim}.signed" -s "${shim_sign_key[@]}"
+  mv "${shim}.signed" "${shim}"
+  rm "${shim}.unsigned"
+  pesign -i "${shim}" -l
+  pesigcheck -i "${shim}" -n 0 -c "${SBKEYS}/db.cer"
```

Comment:
This block of code is repeated at least 3 times (possibly a 4th in the `vmlinuz_sbsetup` function. Can it be turned into its own function?

Reply:
They are close, but different enough where it wouldn't be as clean as you'd think. It could be split into a `remove_pesig` and `add_pesign` but might end up being about the same number of lines the end. I don't mind adding it though.